File: week4/solver.py
import sys
from board import Board
from queue import PriorityQueue

class Solver():
    def __init__(self, initial: Board):
        if not initial:
            raise ValueError("Initial board is None")
        self.initial = initial
        self._moves = 0
        self._solution = []
        self.pq = PriorityQueue()
        self.came_from = {}
        self.solve()

    def solve(self):
        self.pq.put((0, self.initial))
        self.came_from[self.initial] = None  # Начальная доска не имеет предшественника
        paths_costs = {self.initial: 0}

        while not self.pq.empty():
            current_priority, current_board = self.pq.get()
            current_cost = paths_costs[current_board]

            if current_board.is_goal():
                self._moves = current_cost
                self._reconstruct_path(current_board)
                return

            for neighbor in current_board.neighbors():
                next_cost = current_cost + 1
                if neighbor not in paths_costs or next_cost < paths_costs[neighbor]:
                    paths_costs[neighbor] = next_cost
                    priority = next_cost + neighbor.manhattan()
                    self.pq.put((priority, neighbor))
                    self.came_from[neighbor] = current_board

    def _reconstruct_path(self, current_board):
        # Восстановление пути от конечной доски до начальной
        path = []
        while current_board:
            path.append(current_board)
            current_board = self.came_from[current_board]
        self._solution = path[::-1]  # Переворачиваем путь, чтобы начать с начальной доски

          


    def is_solvable(self):
        n = self.initial.dimension()
        tiles = self.initial.tiles
        inversions = 0
        for i in range(n * n - 1):
            for j in range(i + 1, n * n):
                if tiles[i // n][i % n] != 0 and tiles[j // n][j % n] != 0 and tiles[i // n][i % n] > tiles[j // n][j % n]:
                    inversions += 1
        if n % 2 == 0:
            inversions += self.initial.tiles[-1][-1] // n + self.initial.tiles[-1][-1] % n
        return inversions % 2 == 0

    def moves(self):
        if not self.is_solvable():
            return -1
        return self._moves
        

    def solution(self):
        if not self.is_solvable():
            return None
        return self._solution

def main():
    tiles = [[0, 1, 3], [4, 2, 5], [7, 8, 6]]

    initial = Board(tiles)
    solver = Solver(initial)

    if not solver.is_solvable():
        print("No solution possible")
    else:
        print("Minimum number of moves =", solver.moves())
        for board in solver.solution():
            print(board)

if __name__ == "__main__":
    main()

File: JavaFiles/week4/Solver.java
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.MinPQ;
import edu.princeton.cs.algs4.Stack;
import edu.princeton.cs.algs4.StdOut;

public class Solver {

    private boolean isSolvable;

    private int moves;

    private SearchBoard solutionBoard;

    private class SearchBoard implements Comparable<SearchBoard>
    {
        private Board board;

        private SearchBoard previousBoard;

        private int moves;

        private int priority;

        public SearchBoard(Board board, int moves, SearchBoard previousBoard) {
            this.board = board;
            this.moves = moves;
            this.previousBoard = previousBoard;

            priority = board.manhattan() + this.moves;
        }

        public int compareTo(SearchBoard that) {
            if (this.priority > that.priority) {
                return 1;
            }

            if (this.priority < that.priority) {
                return -1;
            }

            return 0;
        }
    }

    public Solver(Board initial)
    {
        MinPQ<SearchBoard> searchBoardQueue = new MinPQ<>();
        MinPQ<SearchBoard> searchTwinBoardQueue = new MinPQ<>();

        searchBoardQueue.insert(new SearchBoard(initial, 0, null));
        searchTwinBoardQueue.insert(new SearchBoard(initial.twin(), 0, null));

        while (true) {
            SearchBoard minBoard = searchBoardQueue.delMin();
            SearchBoard minTwinBoard = searchTwinBoardQueue.delMin();

            if (minBoard.board.isGoal()) {
                solutionBoard = minBoard;
                isSolvable = true;
                moves = minBoard.moves;
                break;
            }

            if (minTwinBoard.board.isGoal()) {
                isSolvable = false;
                moves = -1;
                break;
            }

            enqueueNeighbors(minBoard, searchBoardQueue);
            enqueueNeighbors(minTwinBoard, searchTwinBoardQueue);
        }
    }

    public boolean isSolvable()
    {
        return isSolvable;
    }

    public int moves()
    {
        return moves;
    }

    public Iterable<Board> solution()
    {
        if (!isSolvable) {
            return null;
        }

        Stack<Board> solution = new Stack<>();
        SearchBoard board = solutionBoard;
        while (board != null) {
            solution.push(board.board);
            board = board.previousBoard;
        }

        return solution;
    }

    private void enqueueNeighbors(SearchBoard searchBoard, MinPQ<SearchBoard> queue)
    {
        for (Board nextBoard: searchBoard.board.neighbors()) {
            if ((searchBoard.previousBoard == null) || (!nextBoard.equals(searchBoard.previousBoard.board))) {
                queue.insert(new SearchBoard(nextBoard, searchBoard.moves + 1, searchBoard));
            }
        }
    }

    public static void main(String[] args) {
        // create initial board from file
        In in = new In(args[0]);
        int n = in.readInt();
        int[][] tiles = new int[n][n];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                tiles[i][j] = in.readInt();
        Board initial = new Board(tiles);

        // solve the puzzle
        Solver solver = new Solver(initial);

        // print solution to standard output
        if (!solver.isSolvable())
            StdOut.println("No solution possible");
        else {
            StdOut.println("Minimum number of moves = " + solver.moves());
            for (Board board : solver.solution())
                StdOut.println(board);
        }
    }
}
